File: 08-sparse-matrix/pascal/dynamic_sparse_matrix.pas
program DynamicSparseMatrix;

{$mode objfpc}

uses
    SysUtils;

type
    TNodoC = class
    public
        derecha: TNodoC;
        abajo: TNodoC;
        valorStr: string;
        constructor Create(const v: string);
    end;

    TNodoFila = class
    public
        numFilaVal: Integer;
        siguiente: TNodoFila;
        derecha: TNodoC;
        constructor Create(fila: Integer);
    end;

    TNodoColumna = class
    public
        numColVal: Char;
        siguiente: TNodoColumna;
        abajo: TNodoC;
        constructor Create(col: Char);
    end;

    TDynamicSparseMatrix = class
    public
        primeroFila: TNodoFila;
        primeroColumna: TNodoColumna;
        constructor Create;
        procedure agregarNodoFila(fila: Integer);
        procedure agregarNodoColumna(col: Char);
        procedure agregar(valor: string; columna: Char; fila: Integer);
        procedure mostrarColumnas;
        procedure mostrarFilas;
    end;

constructor TNodoC.Create(const v: string);
begin
    valorStr := v;
    derecha := nil;
    abajo := nil;
end;

constructor TNodoFila.Create(fila: Integer);
begin
    numFilaVal := fila;
    siguiente := nil;
    derecha := nil;
end;

constructor TNodoColumna.Create(col: Char);
begin
    numColVal := col;
    siguiente := nil;
    abajo := nil;
end;

constructor TDynamicSparseMatrix.Create;
begin
    primeroFila := nil;
    primeroColumna := nil;
end;

procedure TDynamicSparseMatrix.agregarNodoFila(fila: Integer);
var
    nuevo: TNodoFila;
begin
    nuevo := TNodoFila.Create(fila);
    if primeroFila = nil then
        primeroFila := nuevo
    else
    begin
        nuevo.siguiente := primeroFila;
        primeroFila := nuevo;
    end;
end;

procedure TDynamicSparseMatrix.agregarNodoColumna(col: Char);
var
    nuevo: TNodoColumna;
begin
    nuevo := TNodoColumna.Create(col);
    if primeroColumna = nil then
        primeroColumna := nuevo
    else
    begin
        nuevo.siguiente := primeroColumna;
        primeroColumna := nuevo;
    end;
end;

procedure TDynamicSparseMatrix.agregar(valor: string; columna: Char; fila: Integer);
var
    nuevo: TNodoC;
    tempCol: TNodoColumna;
    tempFila: TNodoFila;
    temp: TNodoC;
begin
    nuevo := TNodoC.Create(valor);

    tempCol := primeroColumna;
    while tempCol <> nil do
    begin
        if tempCol.numColVal = columna then
        begin
            if tempCol.abajo = nil then
                tempCol.abajo := nuevo
            else
            begin
                temp := tempCol.abajo;
                while temp.abajo <> nil do
                    temp := temp.abajo;
                temp.abajo := nuevo;
            end;
            Break;
        end;
        tempCol := tempCol.siguiente;
    end;

    tempFila := primeroFila;
    while tempFila <> nil do
    begin
        if tempFila.numFilaVal = fila then
        begin
            if tempFila.derecha = nil then
                tempFila.derecha := nuevo
            else
            begin
                temp := tempFila.derecha;
                while temp.derecha <> nil do
                    temp := temp.derecha;
                temp.derecha := nuevo;
            end;
            Break;
        end;
        tempFila := tempFila.siguiente;
    end;
end;

procedure TDynamicSparseMatrix.mostrarColumnas;
var
    temp: TNodoColumna;
begin
    Write('  ');
    temp := primeroColumna;
    while temp <> nil do
    begin
        Write(temp.numColVal, ' ');
        temp := temp.siguiente;
    end;
    Writeln;
end;

procedure TDynamicSparseMatrix.mostrarFilas;
var
    temp: TNodoFila;
    t: TNodoC;
begin
    temp := primeroFila;
    while temp <> nil do
    begin
        Write(temp.numFilaVal, ' ');
        t := temp.derecha;
        while t <> nil do
        begin
            Write(t.valorStr, ' ');
            t := t.derecha;
        end;
        Writeln;
        temp := temp.siguiente;
    end;
end;

var
    sm: TDynamicSparseMatrix;

begin
    sm := TDynamicSparseMatrix.Create;
    sm.agregarNodoColumna('C');
    sm.agregarNodoColumna('B');
    sm.agregarNodoColumna('A');
    sm.agregarNodoFila(3);
    sm.agregarNodoFila(2);
    sm.agregarNodoFila(1);
    sm.agregar('5', 'A', 1);
    sm.agregar('8', 'A', 2);
    sm.agregar('7', 'B', 2);
    sm.agregar('1', 'A', 3);
    sm.agregar('3', 'B', 3);
    sm.agregar('9', 'C', 3);
    sm.mostrarColumnas;
    sm.mostrarFilas;
end.

File: 08-sparse-matrix/cpp/DynamicSparseMatrix.cpp
// Luis Espino 2024

#include <iostream>

using namespace std;

class NodoC{
public:
	NodoC *derecha;
	NodoC *abajo;
	string valor;
	NodoC(string valor){
		this->valor = valor;
		derecha = NULL;
		abajo = NULL;
	}
};

class NodoFila{
public:
	int numFila;
	NodoFila *siguiente;
	NodoC *derecha;
	NodoFila(int numFila){
		this->numFila = numFila;
		siguiente = NULL;
		derecha = NULL;
	}
};

class NodoColumna{
public:
	char numColumna;
	NodoColumna *siguiente;
	NodoC *abajo;
	NodoColumna(char numColumna){
		this->numColumna = numColumna;
		siguiente = NULL;
		abajo = NULL;
	}
};

class DynamicSparseMatrix {
public:
	NodoFila *primeroFila;
	NodoColumna *primeroColumna;

	void agregarNodoFila(int numFila) {
		if (primeroFila==NULL)
			primeroFila = new NodoFila(numFila);
		else {
			NodoFila *temp = primeroFila;
			primeroFila = new NodoFila(numFila);
			primeroFila->siguiente = temp;
		}
	}

	void agregarNodoColumna(char numColumna) {
		if (primeroColumna==NULL)
			primeroColumna = new NodoColumna(numColumna);
		else {
			NodoColumna *temp = primeroColumna;
			primeroColumna = new NodoColumna(numColumna);
			primeroColumna->siguiente = temp;
		}
	}

	void agregar(string valor, char columna, int fila){
		NodoColumna *tempColumna = primeroColumna;
		NodoFila *tempFila = primeroFila;

		NodoC *nuevo = new NodoC(valor);

		while (tempColumna!=NULL) {
			if (tempColumna->numColumna==columna) {
				NodoC *temp = tempColumna->abajo;
				if (temp==NULL)
					tempColumna->abajo = nuevo;
				else {
					while (temp->abajo!=NULL) temp = temp->abajo;
					temp->abajo = nuevo;

				}
				break;
			}
			tempColumna = tempColumna->siguiente;
		}

		while (tempFila!=NULL) {
			if (tempFila->numFila==fila) {
				NodoC *temp = tempFila->derecha;
				if (temp==NULL)
					tempFila->derecha = nuevo;
				else {
					while (temp->derecha!=NULL) temp = temp->derecha;
					temp->derecha = nuevo;
				}
				break;
			}
			tempFila = tempFila->siguiente;
		}
	}

	void mostrarColumnas() {
		NodoColumna *temp = primeroColumna;
		cout << "  ";
		while(temp!=NULL) {
			cout << temp->numColumna << " ";
			temp = temp->siguiente;
		}
		cout << endl;
	}

	void mostrarFilas() {
		NodoFila *temp = primeroFila;
		while(temp!=NULL) {
			cout << temp->numFila << " ";
			NodoC *t = temp->derecha;
			while (t!=NULL){
				cout << t->valor << " ";
				t = t->derecha;
			}
			cout << endl;
			temp = temp->siguiente;
		}
		cout << endl;
	}
};


int main()
{
    DynamicSparseMatrix *sm = new DynamicSparseMatrix();
	sm->agregarNodoColumna('C');
	sm->agregarNodoColumna('B');
	sm->agregarNodoColumna('A');
	sm->agregarNodoFila(3);
	sm->agregarNodoFila(2);
	sm->agregarNodoFila(1);
	sm->agregar("5",'A', 1);
	sm->agregar("8",'A', 2);
	sm->agregar("7",'B', 2);
	sm->agregar("1",'A', 3);
	sm->agregar("3",'B', 3);
	sm->agregar("9",'C', 3);
	sm->mostrarColumnas();
	sm->mostrarFilas();
}